torchvision/models/optical_flow/detector.py

```diff
@@ -0,0 +1,465 @@
+
+import numpy as np
+import matplotlib.pyplot as plt
+from tqdm.notebook import tqdm as tqdm_notebook
+from tqdm import tqdm
+
+import torch
+import torchvision
+import torch.nn as nn
+import torchvision.transforms.functional as F
+import torchvision.transforms as T
+from torch import Tensor
+from torchvision.utils import flow_to_image
+from torchvision.io import write_jpeg, write_png
+
+
+def _release_memory(with_no_grad: bool=True):
+    """Releases memory. 
+    
+    Steps:
+        - runs garbage collection: ``gc.collect()``
+        - empties the gpu cache: ``torch.cuda.empty_cache()``
+    """
+    gc.collect()
+    if torch.cuda.is_available():
+        if with_no_grad:
+            with torch.no_grad():
+                torch.cuda.empty_cache()
+        else:
+            torch.cuda.empty_cache()
+
+
+@torch.no_grad()
+def raft_preprocess(batch: Tensor) -> Tensor:
+    """Returns RAFT model preprocessing transformation."""
+    transforms = T.Compose(
+        [
+            T.ConvertImageDtype(torch.float32),
+            T.Normalize(mean=0.5, std=0.5),  # map [0, 1] into [-1, 1]
+            T.Resize(size=(520, 960)),
+        ]
+    )
+    return transforms(batch)
+
+
+@torch.no_grad()
+def _prepare_frame(frame: Union[Tensor, List[Tensor]], preprocess: Optional[Callable]=None) -> Tensor:
+    """Prepares a frame and returns the preprocessed frame as a Tensor of shape 
+    (N, 3, H, W).
+
+    Args:
+        frame (Union[Tensor, List[Tensor]]): Input frame(s).
+        .. warning::
+            Input ``frame`` should be of shape ``(N, 3, H, W)``, ``(N, 1, H, W)``, 
+            ``(3, H, W)``, ``(1, H, W)`` or ``(H, W)``.
+        preprocess (Callable, optional): Preprocessing function for the model. 
+            By default, this uses the preprocessing function for ``RAFT`` model.
+
+    Returns:
+        frame (Tensor): Preprocessed frame of shape (N, 3, H, W).
+        
+    """
+    frame_shape = frame.shape
+    if isinstance(frame, list):
+        device = frame[0].device
+        frame = torch.stack(frame).to(device)
+    else:
+        device = frame.device
+    # Ensure the frame has the shape: (N, C, H, W); C = 3
+    if frame.dim() == 3:
+        if frame.shape[0] == 1:
+            frame = torch.tile(frame, (3, 1, 1))
+        # Set shape: (1, 3, H, W); N = 1, C = 3
+        frame = frame.unsqueeze(0)
+    elif frame.dim() == 2:
+        # Set shape: (1, 3, H, W); N = 1, C = 3
+        frame = torch.tile(frame.unsqueeze(0), (3, 1, 1)).unsqueeze(0)
+    # Assuming C == 3
+    if frame.shape[1] != 3 or frame.dim() > 4:
+        raise ValueError(dedent(
+            f"""Input should be of shape (N, 3, H, W), (N, 1, H, W), 
+            (3, H, W), (1, H, W) or (H, W); got {frame_shape} 
+            and adjusted to {frame.shape}.
+            """))
+    if preprocess is None:
+        preprocess = raft_preprocess
+    return preprocess(frame).to(device)
+
+
+def detect_flow(frameA: Union[Tensor, List[Tensor]], 
```

Comment:
Hi! @NicolasHug will have a deeper look.

I think that we do lot of implicit assumptions.

Reply:
@oke-aditya About the the assumptions on folder name, saving etc., it was necessary to save the files to disk. Now you may argue (*as I debated with myself*) that it should be perfectly fine returning just the `flow_img` and `predicted_flow` and let the user handle saving the files to disk. However, given the memory constraints, it might be useful, if you could just conveniently specify whether you want to save the raw and processed flow data to disk, or you want them returned as well.

Given, you have this function, `detect_flow` it all boils down to a single function call for generating optical flows. On the other hand, `detect_flows` handles the same for multiple images, which you can now create a video/gif file from.

So, yes, I had envisioned them as parts of `utils/`. But, just to begin with, pushed the initial commit to `torchvision/models/optical_flow/`.